### PR TITLE
add LSH to contrib.tracking

### PR DESCRIPTION
**Pull request by alicanb**
Addresses #1185 

This PR adds LSH from pyro-iei to contrib.tracking. One design choice I made is to allow overwriting in `lsh.add`. We can easily change that to not allow overwriting in `lsh.add` and add `lsh.modify` to facilitate that need.

### CODE REVIEW

**Comment by alicanb** [comment]
BTW these tests take much longer time than I expected, probably mainly in checking set equivalence. Is there a faster way to do that? I don't want to add 6 secs to test time just for this.

File: tests/contrib/tracking/test_hashing.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+def test_lsh_add(scale):
+    lsh = LSH(scale)
+    a = torch.rand(10)
+    lsh.add('a', a)
+    assert lsh._hash_to_key[lsh._key_to_hash['a']] == {'a'}
+
+
+@pytest.mark.parametrize('scale', [0.1, 1, 10, 100])
+def test_lsh_hash_nearby(scale):
+    k = 10
```
Comment (fritzo) [inline tests/contrib/tracking/test_hashing.py]: Lookup has complexity `O(3^k)`. If these tests are slow, try reducing k to say 7.

Comment (fritzo) [reply tests/contrib/tracking/test_hashing.py]: BTW we could easily replace this implementation with an `O(2^k)` version which is much cheaper for higher dimensions.

File: pyro/contrib/tracking/hashing.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import itertools
+from collections import defaultdict
+from numbers import Number
+
+
+class LSH(object):
+    """
+    Implements locality-sensitive hash for low-dimensional euclidean space.
+    """
+    def __init__(self, scale):
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: Can we rename `scale` to `radius` and describe the radius guarantee in the docstring?

File: pyro/contrib/tracking/hashing.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import, division, print_function
+
+import itertools
+from collections import defaultdict
+from numbers import Number
+
+
+class LSH(object):
+    """
+    Implements locality-sensitive hash for low-dimensional euclidean space.
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: Can you describe the `:param float radius: ...` in this docstring, and maybe add an `Example::` section with usage?

File: pyro/contrib/tracking/hashing.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            "scale must be greater than 0, given: {}".format(scale)
+        self._scale = scale
+        self._hash_to_key = defaultdict(set)
+        self._key_to_hash = {}
+
+    def _hash(self, point):
+        coords = (point.detach().cpu() / self._scale).round()
+        return tuple(map(int, coords))
+
+    def add(self, key, point):
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: add docstring for public methods

File: pyro/contrib/tracking/hashing.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        _hash = self._hash(point)
+        if key in self._key_to_hash.keys():
+            self.remove(key)
+        self._key_to_hash[key] = _hash
+        self._hash_to_key[_hash].add(key)
+
+    def remove(self, key):
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: add docstring for public methods

File: pyro/contrib/tracking/hashing.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        _hash = self._key_to_hash.pop(key)
+        self._hash_to_key[_hash].remove(key)
+
+    def nearby(self, key):
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: add docstring for public methods

File: pyro/contrib/tracking/hashing.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        self._bins = set()
+
+    def _hash(self, point):
+        coords = (point.detach().cpu() / self._scale).round()
+        return tuple(map(int, coords))
+
+    def try_add(self, point):
+        """
+        If bin is unoccupied, adds to bin and returns True.
+        If bin is already occupied, returns False.
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: improve docstring: add a `:param torch.Tensor point: ...` section and a `:return: ...` section

File: pyro/contrib/tracking/hashing.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            result |= self._hash_to_key[nearby_hash]
+        result.remove(key)
+        return result
+
+
+class ApproxSet(object):
+    """
+    Queries low-dimensional euclidean space for approximate occupancy.
+    """
+    def __init__(self, scale):
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: Can we rename `scale` to `radius` and describe the guarantee?

File: pyro/contrib/tracking/hashing.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        return tuple(map(int, coords))
+
+    def add(self, key, point):
+        """
+        Adds (`key`, `hash(point)`) pair to the hash
+        :param key: key used identify point
+        :param point: data, should be hashable
+        """
+        _hash = self._hash(point)
+        if key in self._key_to_hash.keys():
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: Remove the unnecessary `.keys()`. In Python2 this creates an entire copy of keys list each call.

**Comment by fritzo** [comment]
Can you ensure that `make docs` works for you locally and produces desired results? On my mac I usually `make docs && open docs/build/html/index.html` and snoop around. A couple style notes on documentation:
- double ticks trigger code font (single ticks are merely italic)
- It's nice to follow a standard pattern separated by double newlines:
    - a single sentence description
    - paragraphs explaining details
    - a :param: section

File: pyro/contrib/tracking/hashing.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _hash(self, point):
+        coords = (point.detach().cpu() / self._radius).round()
+        return tuple(map(int, coords))
+
+    def add(self, key, point):
+        """
+        Adds (`key`, `hash(point)`) pair to the hash
+        :param key: key used identify point
+        :param point: data, should be hashable
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: Isn't it `key` rather than `point` that should be hashable?
Also: `:param torch.Tensor point`

File: pyro/contrib/tracking/hashing.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    """
+    def __init__(self, radius):
+        assert radius > 0 if isinstance(radius, Number) else (radius > 0).all(), \
+            "radius must be greater than 0, given: {}".format(radius)
+        self._radius = radius
+        self._hash_to_key = defaultdict(set)
+        self._key_to_hash = {}
+
+    def _hash(self, point):
+        coords = (point.detach().cpu() / self._radius).round()
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: I think it's reasonable to expect the caller to have called `.detach().cpu()` on the large tensor of points from which this point is sliced out.

**Comment by alicanb** [comment]
Thanks @fritzo, I fixed the docs. something's wrong with the CI build though...

File: pyro/contrib/tracking/hashing.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        >>> a = torch.tensor([-0.51, -0.51]) # hash(a)=(-1,-1)
+        >>> b = torch.tensor([-0.49, -0.49]) # hash(a)=(0,0)
+        >>> c = torch.tensor([1.0, 1.0]) # hash(b)=(1,1)
+        >>> lsh.add('a', a)
+        >>> lsh.add('b', b)
+        >>> lsh.add('c', c)
+        >>> print(lsh.nearby('a')) # {'b'}, even though c is within 2radius of a
+        >>> print(lsh.nearby('b')) # {'a','c'}
+        >>> lsh.remove('b')
+        >>> print(lsh.nearby('a')) # {}
```
Comment (neerajprad) [inline pyro/contrib/tracking/hashing.py]: @alicanb - This is failing the doctest. You can run `make docs` to replicate the failure. You can just remove the prints and append the output below for the tests to pass. Let me know if you need any help fixing this. e.g.

```
>>> lsh.nearby('a')
set(['b'])
```

Comment (alicanb) [reply pyro/contrib/tracking/hashing.py]: Thanks @neerajprad , weird- I can build successfully on my machine... Anyways I pushed a fix, I'll ask you if this doesn't work.

**Comment by fritzo** [comment]
@ae-foster This is a simple low-dimensional euclidean binning algorithm rather than a high-dimensional random hyperplane cut algorithm. Do you think 'LSH' is an appropriate name, or should we rename this to make room for / avoid confusion with a random hyperplane implementation?

**Comment by ae-foster** [comment]
Though I'm no expert, I say LSH goes with any hashing algorithm. E.g. in this package https://github.com/ekzhu/datasketch we called it MinHashLSH- it's LSH using MinHash. You could call it <HashAlgorithm>LSH?

**Comment by neerajprad** [comment]
+1. Maybe `EuclideanLSH` or `EuclideanBinningLSH`?  (leaving room for RandomHyperplanesLSH for cosine distance)?

File: tests/contrib/tracking/test_hashing.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+import pytest
+import torch
+
+from pyro.contrib.tracking.hashing import LSH, ApproxSet
+from tests.common import assert_equal
+
+
+@pytest.mark.parametrize('scale', [-1., 0., -1 * torch.ones(2, 2)])
+def test_lsh_init(scale):
+    num_error = 0
```
Comment (fritzo) [inline tests/contrib/tracking/test_hashing.py]: A cleaner way to write death tests is to use pytest:
```py
@pytest.mark.parametrize('scale', [-1., 0., -1 * torch.ones(2, 2)])
def test_lsh_init(scale):
    with pytest.raises(ValueError):
        LSH(scale)    
```

File: pyro/contrib/tracking/hashing.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+class ApproxSet(object):
+    """
+    Queries low-dimensional euclidean space for approximate occupancy.
+
+
+    :param float radius: scaling parameter used in hash function. Determines the size of the bin.
+                         See :class:`LSH` for details.
+    """
+    def __init__(self, radius):
+        assert radius > 0 if isinstance(radius, Number) else (radius > 0).all(), \
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: int: user-facing errors should be `ValueError`s rather than assertions:
```py
if not (radius > 0):
    raise ValueError("radius must be greater than 0, given: {}".format(radius))
```
Also I think it's fine to require float rather than vector radius. If a user wants a diagonal scale, they can map the `point`s by that scale.

File: pyro/contrib/tracking/hashing.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
+from __future__ import absolute_import, division, print_function
+
+import itertools
+from collections import defaultdict
+from numbers import Number
+
+
+class LSH(object):
+    """
+    Implements locality-sensitive hashing.
```
Comment (fritzo) [inline pyro/contrib/tracking/hashing.py]: How about we keep the class name `LSH` and describe a little more specifically:
> Implements locality-sensitive hashing for low-dimensional euclidean space.

Comment (alicanb) [reply pyro/contrib/tracking/hashing.py]: I kept the name `LSH` for now, will we have additional hashing schemes in the future?

Comment (fritzo) [reply pyro/contrib/tracking/hashing.py]: It's conceivable that we might track in embedding space rather than physical space, and in that case a cosine-distance random-hyperplane LSH would work better than this version. But let's address that when it arises.
